### PR TITLE
Fix issue 18063 - thread_attachThis returns dangling pointer

### PR DESCRIPTION
**Pull request by acehreli**
Prevent returning dangling pointer as well as call all detach varieties after locking slock. (All other calls to Thread.remove() are done while holding slock.)

### CODE REVIEW

File: src/core/thread.d [hunk 1]
```diff
@@ -2239,8 +2244,14 @@ version( Windows )
  */
 extern (C) void thread_detachThis() nothrow @nogc
 {
+    Thread.slock.lock_nothrow();
+    scope(exit) Thread.slock.unlock_nothrow();
```
Comment (nemanja-boric-sociomantic) [inline src/core/thread.d]: Ideally, these changes should be in the separate commit.

Comment (acehreli) [reply src/core/thread.d]: Agreed but I could not pass my test program without fixing these as well. Perhaps I should open two bugs and submit fixes for two.

Comment (nemanja-boric-sociomantic) [reply src/core/thread.d]: Why not just add extra commit in this PR? Make the first commit to add locks, and the second commit (which includes `Fixes issue ...` entry) with `t.setThis(null)` and doc change?

File: test/thread/src/attach_detach.d [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+{
+    foreach (ref t; threads)
+        start_thread(t);
+}
+
+void stop_thread(ref MyThread t)
+{
+    t.stopRequested = true;
+    while (!t.stopped)
+        Thread.sleep(1000.msecs);
```
Comment (acehreli) [inline test/thread/src/attach_detach.d]: I hadn't realized that this sleep was so long. Reducing it to 1.mecs causes a segfault. Working on it...

**Comment by acehreli** [comment]
Closing for now as I'm not sure whether SIGSEGV during pthread_detach() inside code/thread.d is a druntime or a test code bug.
